File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionJsonPatch.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;
import co.nlighten.jsontransform.manipulation.JsonPatch;

import java.util.Map;

public class TransformerFunctionJsonPatch extends TransformerFunction {

    public TransformerFunctionJsonPatch() {
        super(FunctionDescription.of(
                Map.of(
                        "ops", ArgumentType.of(ArgType.Array).position(0)
                )
        ));
    }

    @Override
    public CompletionStage<Object> apply(FunctionContext context) {
        var source = context.getJsonElement(null);
        if (source == null) {
            return null;
        }
        return new JsonPatch(context.getAdapter()).patch(context.getJsonArray("ops"), source);
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionJsonPatch.ts
import JsonPatch from "fast-json-patch";
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";

class TransformerFunctionJsonPatch extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        ops: { type: ArgType.Array, position: 0 },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const source = await context.getJsonElement(null);
    if (source == null) {
      return null;
    }
    const ops = (await context.getJsonArray("ops")) ?? [];
    return JsonPatch.applyPatch(source, ops, true).newDocument;
  }
}

export default TransformerFunctionJsonPatch;
